### DISCUSSION TITLE
Translating Modern Compiler Implementation in ML

### DISCUSSION THREAD

**Discussion by lee-thomp** [coalton-lang/coalton#1221] [Help!] [answered]
How would I go about translating the following from Figure 1.4 of "Modern Compiler Implementation in ML" into Coalton? I'm stuck on the mutual recursion of `stm` and `exp`:
```sml
type id = string
datatype binop = Plus | Minus | Times | Div
datatype stm = CompoundStm of stm * stm
             | AssignStm of id * exp
             | PrintStm of exp list
     and exp = IdExp of id
             | NumExp of int
             | OpExp of exp * binop * exp
             | EseqExp of stm * exp

**Comment by stylewarning** [comment, marked as answer]
The only thing that's not directly supported are type aliases. That's a feature being designed right now. So you'd either need to just use a dedicated type for that, or don't use it at all.

In this example, I'll just make a new type called `Ident`. A translation would look something like this.

```
(coalton-toplevel
  (define-type Ident
    (Ident String))

  (define-type BinOp
    Plus
    Minus
    Times
    Div)

  (define-type Stm
    (CompoundStm Stm Stm)
    (AssignStm Ident Exp)
    (PrintStm (List Exp)))

  (define-type Exp
    (IdExp Ident)
    (NumExp Integer)
    (OpExp Exp BinOp Exp)
    (ESeqExp Stm Exp)))
```

I would love to see "Modern Compiler Implementation in Coalton"! 😊

**Comment by lee-thomp** [comment]
Ah okay I found the source of my issue.  I'd actually translated it exactly as you described before posting my message.  I'd managed to catch the issue where `Id` is already defined in Coalton, but not where `Exp` is already defined (`coalton-library/math/elementary`).  Using [Sly](https://joaotavora.github.io/sly/) with Emacs and SBCL showed me a whole backtrace of type check errors but nothing specific about `Exp` itself.  I only caught this after loading this all into a bare SBCL repl where I was told about the math package already having its own `Exp`.

Replacing `Exp` with `Expr` for example fixed all this and typechecks fine.

**Reply by stylewarning** [reply]
Ah great!

I recommend defining your own package and using package-local nicknames. :)
